agent/src/main/java/com/appland/appmap/util/GitUtil.java

```diff
@@ -0,0 +1,237 @@
+    try {
+      FileRepositoryBuilder builder = new FileRepositoryBuilder()
+          .readEnvironment();
+      if (workTree != null) {
+        builder.setWorkTree(workTree);
+      } else {
+        builder.findGitDir();
+      }
+      return new GitUtil(new Git(builder.build()));
+    } catch (IOException e) {
+      logger.warn(e);
+    }
+    return null;
+  }
+
+  @Override
+  public void close() {
+    git.close();
+  }
+
+  public Repository getRepository() {
+    return git.getRepository();
+  }
+
+  public String getRepositoryURL() {
+    try {
+      List<RemoteConfig> remotes = git.remoteList().call();
+      Optional<RemoteConfig> originConfig = remotes.stream().filter(r -> r.getName().equals("origin")).findFirst();
+      List<URIish> uris = originConfig.isPresent() ? originConfig.get().getURIs() : remotes.get(0).getURIs();
+      return uris.get(0).toASCIIString();
+    } catch (GitAPIException e) {
+      logger.warn(e);
+    }
+    return "";
+  }
+
+  public String getBranch() {
+    try {
+      return getRepository().getBranch();
+    } catch (IOException e) {
+      logger.warn(e);
+    }
+    return "";
+  }
+
+  public String getCommit() {
+    try {
+      return getRepository().resolve(HEAD).name();
+    } catch (RevisionSyntaxException | IOException e) {
+      logger.warn(e);
+    }
+    return "";
+  }
+
+  public List<String> getStatus() {
+    Status status;
+    List<String> ret = new ArrayList<>();
+    try {
+      status = git.status().call();
+      ret.addAll(status.getAdded().stream().map(f -> "A  " + f).collect(Collectors.toList()));
+      ret.addAll(status.getChanged().stream().map(f -> " M " + f).collect(Collectors.toList()));
+      ret.addAll(status.getModified().stream().map(f -> " M " + f).collect(Collectors.toList()));
+      ret.addAll(status.getMissing().stream().map(f -> "D  " + f).collect(Collectors.toList()));
+      ret.addAll(status.getRemoved().stream().map(f -> "D  " + f).collect(Collectors.toList()));
+      ret.addAll(status.getUntracked().stream().map(f -> "?? " + f).collect(Collectors.toList()));
+      return ret;
+    } catch (NoWorkTreeException | GitAPIException e) {
+      logger.warn(e);
+    }
+    return ret;
+  }
+
+  // In the code below, the "fs" prefix indicates that the variable holds a
+  // filesystem path. The "git" prefix indicates that the variable holds a git
+  // path. Hopefully this will make mistakes easier to spot.
+
+  private static final byte[] GIT_MAIN_JAVA = "src/main/java".getBytes();
+  private static int GIT_MAIN_JAVA_LEN = GIT_MAIN_JAVA.length;
+  private static final byte[] GIT_TEST_JAVA = "src/test/java".getBytes();
+  private static int GIT_TEST_JAVA_LEN = GIT_TEST_JAVA.length;
+
+  /**
+   * If we're in a git repo, walk all the directories under the directory that
+   * contains the config, looking for source directories (i.e. those that staart
+   * "src/main/java" or "src/test/java").
```

Comment:
Is it always the case that Java sources are in directories like that? I'm not sure how widespread that convention is. Should we be worried that projects with a different structure would fall through the cracks, or is it so rare that we needn't care?

Reply:
Always? No. However, `src/main/java` and `src/test/java` are the default trees for Java source in maven and gradle projects. As far as I can tell, most projects use them, possibly adding additional directories, if required (e.g. for generated sources).

If the agent can't determine the source's real location, it falls back to the old behavior. So, projects that don't use `src/{main,test}/java` will be no worse off.